Fix auto_mixed_precision handling of data flow ops

benbarsdell opened the pull request:
TensorArray and Stack ops are stateful and therefore not always safe to modify (e.g., they may get shared between training and validation graphs, which requires that their data types remain the same).

This commit removes support for these ops and replaces it with support for TensorList ops, which are non-stateful (and used by default as of TF 2.0).

This may cause performance regressions for models using auto_mixed_precision in TF 1, but this can be worked around by calling tf.enable_control_flow_v2() or switching to TF 2.

Attn. @reedwm 
cc. @nluehr

benbarsdell committed:
Fix auto_mixed_precision handling of data flow ops

- TensorArray and Stack ops are stateful and therefore not always safe
  to modify (e.g., they may get shared between training and validation
  graphs, which requires that their data types remain the same).
- This commit removes support for these ops and replaces it with support
  for TensorList ops, which are non-stateful (and used by default as of
  TF 2.0).
- This may cause performance regressions for models using
  auto_mixed_precision in TF 1, but this can be worked around by
  calling tf.enable_control_flow_v2() or switching to TF 2.

reedwm commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1202,20 +1204,24 @@ Status AutoMixedPrecisionImpl::Optimize() {
-                               &object_clients_map));
+      {"EmptyTensorList", "TensorListSplit", "TensorListFromTensor",
+       "TensorListReserve", "TensorListScatter", "TensorListScatterV2",
+       "TensorListConcatLists"},
```
I don't think TensorListConcatLists is handled correctly. You also have to change the dtypes of the input lists.

Maybe give up trying to handle changing the dtype for that tensor list if you encounter a TensorListConcatLists op. Or add a TODO.

benbarsdell replied:
Thanks for pointing this out. I've worked around this along with TensorListPushBackBatch.

benbarsdell committed:
WAR TensorListPushBackBatch/ConcatLists ops

- These ops are not currently handled by the auto_mixed_precision pass,
  so to avoid breaking the graph we disable all handling of tensor list
  ops if any of these are present.
- Adds a test to ensure these ops don't break the graph.

benbarsdell committed:
Support tensor list ops that act as writers

- Some tensor list ops construct a tensor list object and write to it at
  the same time. This commit adds these ops to the clearlist and adds
  ephemeral edges between them and their corresponding reader ops.
- Also adds a test for this behavior.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1202,20 +1204,24 @@ Status AutoMixedPrecisionImpl::Optimize() {
   DataStructureOpsMap object_clients_map;
-  VLOG(2) << "Identifying Stack*V2 nodes";
+
+  // TODO(benbarsdell): Work out how to handle "TensorListPushBackBatch".
```
Thanks for pointing this out. I've added a workaround that will prevent it from breaking the graph.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1202,20 +1204,24 @@ Status AutoMixedPrecisionImpl::Optimize() {
+       {"TensorListGather", TypeAttrId("element_dtype")}},
+      &object_clients_map));
 
   // Create ephemeral edges between writers and readers of data structure ops.
```
That's right. The forcing of ops to have the same dtype is handled independently of the ephemeral edges.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1202,20 +1204,24 @@ Status AutoMixedPrecisionImpl::Optimize() {
-                                {"TensorArrayConcatV3", TypeAttrId("dtype")}},
-                               &object_clients_map));
+      {"EmptyTensorList", "TensorListSplit", "TensorListFromTensor",
+       "TensorListReserve", "TensorListScatter", "TensorListScatterV2",
```
Thanks for pointing this out. I've added support for this behavior.

benbarsdell commented:
Sorry for the delay in getting back to this, I had to think about it for a while. Your feedback was very helpful, thanks. I decided to just work around the PushBackBatch and ConcatLists ops for now, as it will probably require some refactoring to support them properly (and as far as I can see they're not actually used anywhere in TF).

One thing I'm not sure about is what the semantics of read after write dependencies are with these tensor list ops. It may not matter, but I wondered if we should be taking it into account with the ephemeral edges somehow. I couldn't see any documentation on exactly how these ops behave.

reedwm reviewed (changes requested):
Thanks for the changes! Dealing with tensor lists (or any variant type with an internal dtype) is difficult but you are making it work. I added one comment.

As for read-after-write semantics, it works just like any other op. No tensor list op mutates any state. Write ops take in a tensor list and return a new tensor list, leaving the old one unmodified. A subsequent read op should take in the return value of the write op, so it can read the new value. No control dependencies are needed with tensor list ops, since they are not stateful. You have the semantics correct when handling tensor lists in this PR.

benbarsdell committed:
Handle tensor list ops more robustly in auto_mixed_precision

- This will now catch any new ops that start with "TensorList", which
  would otherwise break the graph.
- Also added a warning message when unsupported tensor list ops are
  found in the graph.

reedwm commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1203,16 +1203,39 @@ Status AutoMixedPrecisionImpl::Optimize() {
+      "TensorListConcat",
+      "TensorListConcatV2",
+      "TensorListGetItem",
+      "TensorListGather"};
```
TensorListLength and TensorListElementShape can be added to this list. And you should change the sentence "If an op is added to this list, it should also be added..." to "If an op is added to this list, it should **probably** also be added...".

Also maybe handle TensorListResize?

benbarsdell replied:
Good point.

Does TensorListResize require special handling?

reedwm replied:
You can treat TensorListResize as a read or write op, and everything will do the right thing. It isn't really either, but you need the `ForceColorMatchBetweenDataStructureOps` function to change the output dtype if it changes the input dtype, or vice versa.

You can also just ignore TensorListResize for now and fix it later.

benbarsdell replied:
TensorListResize has no dtype attribute, so it can't be passed to AddDataStructureOpsToMap.

I may be missing something, but won't it just appear as:

SomeTensorListConstructor (A) -> TensorListResize (B) -> SomeTensorListReaderOrWriter (C)

In this case, C will search for its corresponding tensor list constructor (A) by passing straight through B. This will add A and C to the map, and they will get color matched. (I tried this out in a simple test).

If this is correct, then we can just add TensorListResize to supported_list_ops too and it should be good. Is there some way that it could go wrong?

I also notice that TensorListResize is marked as stateful, but I don't understand why?

reedwm replied:
Ah you are right, TensorListResize can just be added to supported_list_ops. 

Good catch on TensorListResize being stateful. This is probably a bug. I'll ask the author.

reedwm replied:
I confirmed TensorListResize being stateful is a bug. Feel free to create a PR fixing this.

benbarsdell committed:
Allow trivial tensor list ops in auto_mixed_precision

- These ops do not have data type attributes and so do not require any
  special handling.
